### PR TITLE
SSHCertificateBuilder

### PR DESCRIPTION
**Pull request by reaperhulk**
This adds support for generating SSH certificates

### CODE REVIEW

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 1]
```diff
@@ -1035,3 +1039,353 @@ def serialize_ssh_public_key(public_key: _SSH_PUBLIC_KEY_TYPES) -> bytes:
...
+            critical_options=self._critical_options,
+            extensions=self._extensions,
+        )
+
+    def add_critical_option(
+        self, name: bytes, value: bytes
+    ) -> "SSHCertificateBuilder":
+        if not isinstance(name, bytes) or not isinstance(value, bytes):
+            raise TypeError("name and value must be bytes")
+        if name in [x[0] for x in self._critical_options]:
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:1237]:
```suggestion
        if name in [name for name, _ in self._critical_options]:
```

and add a comment that its O(n^2)

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 2]
```diff
@@ -1035,3 +1039,353 @@ def serialize_ssh_public_key(public_key: _SSH_PUBLIC_KEY_TYPES) -> bytes:
...
+            critical_options=self._critical_options + [(name, value)],
+            extensions=self._extensions,
+        )
+
+    def add_extension(
+        self, name: bytes, value: bytes
+    ) -> "SSHCertificateBuilder":
+        if not isinstance(name, bytes) or not isinstance(value, bytes):
+            raise TypeError("name and value must be bytes")
+        if name in [x[0] for x in self._extensions]:
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:1257]: same

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 3]
```diff
@@ -1035,3 +1039,353 @@ def serialize_ssh_public_key(public_key: _SSH_PUBLIC_KEY_TYPES) -> bytes:
...
+        f.put_u32(self._type.value)
+        f.put_sshstr(key_id)
+        fprincipals = _FragList()
+        for p in self._valid_principals:
+            fprincipals.put_sshstr(p)
+        f.put_sshstr(fprincipals.tobytes())
+        f.put_u64(_datetime_to_utc_timestamp(self._valid_after))
+        f.put_u64(_datetime_to_utc_timestamp(self._valid_before))
+        fcrit = _FragList()
+        for opt in self._critical_options:
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:1332]: unpacking

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 4]
```diff
@@ -1035,3 +1039,353 @@ def serialize_ssh_public_key(public_key: _SSH_PUBLIC_KEY_TYPES) -> bytes:
...
+            fcrit.put_sshstr(opt[0])
+            fcrit.put_sshstr(opt[1])
+        f.put_sshstr(fcrit.tobytes())
+        fext = _FragList()
+        for ext in self._extensions:
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:1337]: unpacking

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 5]
```diff
@@ -1035,3 +1039,353 @@ def serialize_ssh_public_key(public_key: _SSH_PUBLIC_KEY_TYPES) -> bytes:
 
     pub = binascii.b2a_base64(f_pub.tobytes()).strip()
     return b"".join([key_type, b" ", pub])
+
+
+def _datetime_to_utc_timestamp(time: datetime.datetime) -> int:
+    if time.tzinfo is not None:
+        offset = time.utcoffset()
+        offset = offset if offset else datetime.timedelta()
+        new_time = time.replace(tzinfo=datetime.timezone.utc) - offset
```
[the remark this answers is not in the document]
Comment (reaperhulk) [reply src/cryptography/hazmat/primitives/serialization/ssh.py:1048]: 🤷 -- there's no "treat this naïve datetime as UTC for the purposes of calling `timestamp()` that I could find, but you know I can't read so maybe I'm wrong! This follows the pattern we use for adjusting tz-aware datetimes in x509.

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 6]
```diff
@@ -1035,3 +1039,355 @@ def serialize_ssh_public_key(public_key: _SSH_PUBLIC_KEY_TYPES) -> bytes:
...
+        # that behavior.
+        if self._valid_principals is None:
+            raise ValueError("valid_principals must be set")
+
+        if self._valid_before is None:
+            raise ValueError("valid_before must be set")
+
+        if self._valid_after is None:
+            raise ValueError("valid_after must be set")
+
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:1309]: Check that `valid_after < valid_before`

File: tests/hazmat/primitives/test_ssh.py [hunk 7]
```diff
@@ -1236,5 +1241,330 @@ def test_public_bytes(self, backend):
...
+        assert isinstance(cert_public_key, ec.EllipticCurvePublicKey)
+        assert cert_public_key.public_numbers() == public_key.public_numbers()
+        assert cert.serial == 0
+        assert cert.type is SSHCertificateType.USER
+        assert cert.key_id == key_id
+        assert cert.valid_principals == valid_principals
+        assert cert.valid_before == utc_time_before
+        assert cert.valid_after == valid_after
+        assert cert.critical_options == {b"ordered": b"", b"maybe": b"test2"}
+        assert list(cert.critical_options.keys()) == [b"maybe", b"ordered"]
```
Comment (alex) [inline tests/hazmat/primitives/test_ssh.py:1284]:
```suggestion
        assert list(cert.critical_options) == [b"maybe", b"ordered"]
```

File: tests/hazmat/primitives/test_ssh.py [hunk 8]
```diff
@@ -1236,5 +1241,330 @@ def test_public_bytes(self, backend):
...
+        assert cert.extensions == {b"test": b"a value", b"allowed": b""}
+        assert list(cert.extensions.keys()) == [b"allowed", b"test"]
```
Comment (alex) [inline tests/hazmat/primitives/test_ssh.py:1286]:
```suggestion
        assert list(cert.extensions) == [b"allowed", b"test"]
```

same for all usages of `keys`

**Review by alex — COMMENTED** [review]
Can we get a test or two that checks the exacts bytes of the result?

**Comment by alex** [comment]
Err, wait the nonce makes that hard/impossible. Nevermind.

File: docs/hazmat/primitives/asymmetric/serialization.rst [hunk 9]
```diff
@@ -632,6 +632,125 @@ The format used by OpenSSH for certificates, as specified in
...
+    .. method:: key_id(key_id)
+
+        :param key_id: The key ID to be included in the certificate. This is
+            not a required value.
+        :type key_id: bytes
+
+    .. method:: valid_principals(valid_principals)
+
+        :param valid_principals: A list of principals that the certificate is
+            valid for. This is not a required value, but if not provided the
```
Comment (alex) [inline docs/hazmat/primitives/asymmetric/serialization.rst:705]: This is confusing. We _do_ require you to specify valid principals.

I think what you want to say is: if an empty list is provided, then the certificate is valid for any principal.

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 10]
```diff
@@ -1035,3 +1039,360 @@ def serialize_ssh_public_key(public_key: _SSH_PUBLIC_KEY_TYPES) -> bytes:
...
+            _key_id=key_id,
+            _valid_principals=self._valid_principals,
+            _valid_before=self._valid_before,
+            _valid_after=self._valid_after,
+            _critical_options=self._critical_options,
+            _extensions=self._extensions,
+        )
+
+    def valid_principals(
+        self, valid_principals: typing.List[bytes]
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:1169]: On Nth thought: Can we have require `valid_principals` to have items and add a seperate `valid_for_all_principals()` buidler methods? Seems way less error prone -- and callers must use one or the other.

Comment (reaperhulk) [reply src/cryptography/hazmat/primitives/serialization/ssh.py:1169]: So we'd have two methods for generating but the `valid_principals` attribute on `SSHCertificate` would return an empty list to say it's valid for all principals? We could have it do that, but also have a boolean `valid_for_all_principals` attribute on the cert object I suppose.

Comment (alex) [reply src/cryptography/hazmat/primitives/serialization/ssh.py:1169]: Yeah. I don't love the asymmetry, but its fails closed on both sides:

1. On the signing side, if you accidentally provide an empty list, you don't get all principals
2. On the verifying side, if you don't special case all principals then you get no access

**Review by alex — APPROVED** [review]

**Pull request merged**